clash-prelude/src/Clash/Class/IndexTuple/Deriving.hs

```diff
@@ -0,0 +1,62 @@
+{-|
+Copyright  :  (C) 2023,      QBayLogic B.V.
+License    :  BSD2 (see the file LICENSE)
+Maintainer :  QBayLogic B.V. <devops@qbaylogic.com>
+-}
+
+{-# LANGUAGE TemplateHaskell #-}
+
+module Clash.Class.IndexTuple.Deriving
+  (IndexTuple(..), deriveIndexTupleInstances) where
+
+import Control.Monad.Extra (concatMapM)
+import Language.Haskell.TH
+
+class IndexTuple a where
+  -- | This function can be used to bind variables to numbers indexing
+  -- successive elements in a list-like structure. If the type of the indices
+  -- can be derived from context, it is enough to provide a tuple with the
+  -- variables like:
+  --
+  -- @
+  -- (a, b) = indexTuple
+  -- (a, b, c, d, e, f) = indexTuple
+  -- @
+  --
+  -- and so on. @a@ will be 0, @b@ 1, @c@ 2, etcetera. All variables will have
+  -- the same type, and specifying the type of any one of the variables is
+  -- enough. The type needs to be an instance of `Num`.
+  --
+  -- If a type needs to be specified here, you can use
+  --
+  -- @
+  -- (a :: Int, b, c) = indexTuple
+  -- @
+  --
+  -- or
+  --
+  -- @
+  -- a :: Int
+  -- (a, b, c) = indexTuple
+  -- @
+  --
+  -- depending on preference.
```

Comment:
From this description alone it looks like a slightly different copy of  `indicesI`.
Which it is, but not without reason.
Maybe add something along the lines of:
> This is similar to [`indicesI`](https://hackage.haskell.org/package/clash-prelude-1.6.4/docs/Clash-Sized-Vector.html#v:indicesI) but can be without incurring non-exhaustive pattern warnings with `-Wincomplete-uni-patterns` enabled.
>Note that `-Wincomplete-uni-patterns` is implied by `-Wall` starting from GHC x.y.
> It's intended for use in defining inline primitives, see this example.

Where `this example` is a link to an example of using it in an inline primitive, either in `Clash.Annotations.Primitive` or possible in `Clash.Tutorial`.